### PR TITLE
Add R generator httr2

### LINKED ISSUE #389: R + httr2
It would be great if the converter for R would also support httr2 (https://httr2.r-lib.org/index.html).

### PR DESCRIPTION
**Pull request by mgirlich**
Closes #389.
@hadley can you also have a look at this?

### CODE REVIEW

File: test/fixtures/r-httr2/bash_redirect.r [hunk 1]
```diff
@@ -0,0 +1,21 @@
+library(httr2)
+
+params = list(
```
Comment (hadley) [inline test/fixtures/r-httr2/bash_redirect.r:3]: Why put these in their own variables instead of just inline?

File: test/fixtures/r-httr2/bash_redirect_heredoc.r [hunk 2]
```diff
@@ -0,0 +1,12 @@
+library(httr2)
+
+headers = c(
+  `'Accept'` = "'application/json'",
```
Comment (hadley) [inline test/fixtures/r-httr2/bash_redirect_heredoc.r:4]: Too many quotes?

File: test/fixtures/r-httr2/commented_json_with_session.r [hunk 3]
```diff
@@ -0,0 +1,11 @@
+library(httr2)
+
+headers = c(
+  `Accept` = "application/json"
+)
+
+request("http://localhost:28139") |> 
+  req_method("POST") |> 
+  req_headers(!!!headers) |> 
+  req_body_raw("{   }", "application/json") |>
```
Comment (hadley) [inline test/fixtures/r-httr2/commented_json_with_session.r:10]: I'd make the `type` argument name explicit here.

File: test/fixtures/r-httr2/bash_redirect_heredoc.r (hunk 2 above)
Comment (mgirlich) [reply test/fixtures/r-httr2/bash_redirect_heredoc.r]: This is basically from `curl http://example.com -H "'Accept': 'application/json'"`. So, it is parsed correctly I think. By the way this crashes `httr2::curl_translate()`

File: test/fixtures/r-httr2/bash_redirect.r (hunk 1 above)
Comment (mgirlich) [reply test/fixtures/r-httr2/bash_redirect.r]: Copy-paste from the version for httr. But I inlined them now.

File: test/fixtures/r-httr2/bash_redirect_heredoc.r (hunk 2 above)
Comment (hadley) [reply test/fixtures/r-httr2/bash_redirect_heredoc.r]: Ah got it.

File: test/fixtures/r-httr2/get_charles_syntax.r [hunk 4]
```diff
@@ -0,0 +1,10 @@
+library(httr2)
+
+request("http://localhost:28139/?format=json&") |>
```
Comment (hadley) [inline test/fixtures/r-httr2/get_charles_syntax.r:3]: Why doesn't this use `req_url_query()`? The trailing `&`?

File: test/fixtures/r-httr2/get_proxy.r [hunk 5]
```diff
@@ -0,0 +1,4 @@
+library(httr2)
+
+request("http://localhost:28139") |>
```
Comment (hadley) [inline test/fixtures/r-httr2/get_proxy.r:3]: Could this use `req_proxy()`?

File: test/fixtures/r-httr2/get_max_time_and_connect_timeout.r [hunk 6]
```diff
@@ -0,0 +1,4 @@
+library(httr2)
+
+request("http://localhost:28139") |>
```
Comment (hadley) [inline test/fixtures/r-httr2/get_max_time_and_connect_timeout.r:3]: Could this use `req_timeout`?

File: test/fixtures/r-httr2/get_with_all_numerical_data.r [hunk 7]
```diff
@@ -0,0 +1,10 @@
+library(httr2)
+
+request("http://localhost:28139/CurlToNode") |> 
+  req_method("POST") |> 
+  req_headers(`Accept` = "application/json") |> 
+  req_body_raw(
+    "18233982904",
```
Comment (hadley) [inline test/fixtures/r-httr2/get_with_all_numerical_data.r:7]: Given title of fixture, should this be a number instead of a string?

File: test/fixtures/r-httr2/get_with_env_var.r [hunk 8]
```diff
@@ -0,0 +1,6 @@
+library(httr2)
+
+request("http://localhost:28139/v2/images") |> 
+  req_url_query(`type` = "distribution") |> 
+  req_headers(`Authorization` = paste("Bearer ", Sys.getenv("DO_API_TOKEN"), sep = "")) |>
```
Comment (hadley) [inline test/fixtures/r-httr2/get_with_env_var.r:5]: `paste0()`?

File: test/fixtures/r-httr2/max_redirs.r [hunk 9]
```diff
@@ -0,0 +1,4 @@
+library(httr2)
+
+request("http://localhost:28139") |>
```
Comment (hadley) [inline test/fixtures/r-httr2/max_redirs.r:3]: Could use `req_option()` here?

**Review by hadley — COMMENTED** [review]
This all looks great to me. I pointed out a few places where you could maybe add additional `req_` functions, but this is already a nice improvement 😄

File: src/generators/r/r-httr2.ts [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+export const supportedArgs = new Set([
+  ...COMMON_SUPPORTED_ARGS,
+  "form",
+  "form-string",
+  "insecure",
+  "no-insecure",
+]);
+
+const regexBacktickEscape = /`|\\|\p{C}|[^ \P{Z}]/gu;
+function reprBacktick(s: Word | string): string {
```
Comment (verhovsky) [inline src/generators/r/r-httr2.ts:19]: Import this from the other R generator instead of copying.

File: test/fixtures/r-httr2/get_charles_syntax.r (hunk 4 above)
Comment (mgirlich) [reply test/fixtures/r-httr2/get_charles_syntax.r:3]: The trailing `&` seems to be an issue for the parser.
@verhovsky Why does `curl --compressed "http://localhost:28139/?format=json&"` result in a request with an empty query list but the query array is `format=json&`. Is that on purpose?

Comment (verhovsky) [reply test/fixtures/r-httr2/get_charles_syntax.r:3]: We want to preserve the trailing `&` so that the generated code is actually equivalent to the input curl command.

queryList is empty because most HTTP libraries, if they have a way of passing the query as a list/dictionary, don't have a mechanism to send '' (does httr2?) without a =. If you do something like `params={'': ''}` in Python for example, that would send `format=json&=`. It's easier to catch this in the parser and not have a queryList at all, so that we fallback to requesting a raw URL, which would send the query as-is.

https://github.com/curlconverter/curlconverter/blob/master/src/Query.ts#L82

Comment (mgirlich) [reply test/fixtures/r-httr2/get_charles_syntax.r:3]: Okay, makes sense, thanks!

File: test/fixtures/r-httr2/get_with_all_numerical_data.r (hunk 7 above)
Comment (mgirlich) [reply test/fixtures/r-httr2/get_with_all_numerical_data.r:7]: We could try to detect it but most other parsers also don't. So, I think it's fine to skip this edge case for now.

File: src/generators/r/httr2.ts [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+export function _toRHttr2(
+  requests: Request[],
+  warnings: Warnings = [],
+): string {
+  const request = getFirst(requests, warnings);
+
+  // const cookieList = getCookieList(request, warnings);
+  const queryList = getQueryList(request);
+
+  const contentType = request.headers.getContentType();
```
Comment (verhovsky) [inline src/generators/r/httr2.ts:170]: This is not ideal because it'll remove things like `; charset=UTF-8`

File: src/generators/r/httr2.ts [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+  }
+
+  if (request.data.startsWith("@") && !request.isDataRaw) {
+    const filePath = request.data.slice(1);
+    steps = addCurlStep(steps, "req_body_file", [repr(filePath)]);
+  } else {
+    const [parsedQueryString] = parseQueryString(request.data);
+    const dataIsList = parsedQueryString && parsedQueryString.length;
+    if (dataIsList) {
+      steps = addCurlStep(steps, "req_body_form", [], parsedQueryString);
```
Comment (verhovsky) [inline src/generators/r/httr2.ts:105]: Similarly, this not checking the content-type header is not ideal, since the generated code might end up sending a different content-type.

**Comment by verhovsky** [comment]
Thanks
